src/Build/BackEnd/Components/Scheduler/Scheduler.cs

```diff
@@ -2280,8 +2273,35 @@ private void AssignGlobalRequestId(BuildRequest request)
                 }
             }
 
-            request.GlobalRequestId = _nextGlobalRequestId;
-            _nextGlobalRequestId++;
+            // No matching request found, assign a new ID
+            request.GlobalRequestId = _nextGlobalRequestId++;
+        }
+
+        /// <summary>
+        /// Computes a hash code for a collection of targets, ignoring order and case.
+        /// </summary>
+        private int ComputeTargetsHash(List<string> targets)
+        {
+            int hash = 17; // Start with a non-zero seed
+            foreach (string target in targets)
+            {
+                hash = hash * 31 + StringComparer.OrdinalIgnoreCase.GetHashCode(target);
+            }
+
+            return hash;
+        }
+
+        /// <summary>
+        /// Determines if two target collections contain the same targets, ignoring order and case.
+        /// </summary>
+        private static bool TargetsMatch(HashSet<string> firstTargetsSet, List<string> secondTargetsList)
+        {
+            if (firstTargetsSet.Count != secondTargetsList.Count)
+            {
+                return false;
+            }
+
+            return firstTargetsSet.SetEquals(secondTargetsList);
```

Comment:
I'm leaning towards having our own comparison rather than calling `SetEquals`. The constraints in this case are more restrictive than the general `SetEquals` which lets us be more efficient since we don't care about duplicates. This should cover our case after we ensure the collections are the same size:

    foreach (string target in secondTargetsList)
    {
        if (!firstTargetsSet.Contains(target))
        {
            return false;
        }
    }

    return true;

This avoids:

- Boxing the `List<string>` enumerator
- Some minimal type casting inside of `SetEquals()`
- Allocation of `BitHelper` class in the `CheckUniqueAndUnfoundElements()` (called by `SetEquals()`)

Reply:
applied the change, thanks for the explanation.